src/smt/smt_engine_state.cpp

```diff
@@ -0,0 +1,302 @@
+/*********************                                                        */
+/*! \file smt_engine_state.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Utility for maintaining the state of the SMT engine.
+ **/
+
+#include "smt/smt_engine_state.h"
+
+#include "options/smt_options.h"
+#include "smt/smt_engine.h"
+
+namespace CVC4 {
+namespace smt {
+
+SmtEngineState::SmtEngineState(SmtEngine& smt)
+    : d_smt(smt),
+      d_context(new context::Context),
+      d_userContext(new context::UserContext),
+      d_pendingPops(0),
+      d_fullyInited(false),
+      d_queryMade(false),
+      d_needPostsolve(false),
+      d_status(),
+      d_expectedStatus(),
+      d_smtMode(SmtMode::START)
+{
+}
+
+void SmtEngineState::notifyExpectedStatus(const std::string& status)
+{
+  Assert(status == "sat" || status == "unsat" || status == "unknown")
+      << "SmtEngineState::notifyExpectedStatus: unexpected status string "
+      << status;
+  d_expectedStatus = Result(status, d_filename);
+}
+
+void SmtEngineState::notifyResetAssertions()
+{
+  doPendingPops();
+  while (!d_userLevels.empty())
+  {
+    userPop();
+  }
+  // Remember the global push/pop around everything when beyond Start mode
+  // (see solver execution modes in the SMT-LIB standard)
+  Assert(d_userLevels.size() == 0 && d_userContext->getLevel() == 1);
+  popto(0);
+}
+
+void SmtEngineState::notifyCheckSat(bool hasAssumptions)
+{
+  if (d_queryMade && !options::incrementalSolving())
+  {
+    throw ModalException(
+        "Cannot make multiple queries unless "
+        "incremental solving is enabled "
+        "(try --incremental)");
+  }
+
+  // Note that a query has been made and we are in assert mode
+  d_queryMade = true;
+  d_smtMode = SmtMode::ASSERT;
+
+  // push if there are assumptions
+  if (hasAssumptions)
+  {
+    internalPush();
+  }
+}
+
+void SmtEngineState::notifyCheckSatResult(bool hasAssumptions, Result r)
+{
+  d_needPostsolve = true;
```

Comment:
That comment should be updated, will do.

The SmtEngine is responsible for making the actual calls to TheoryEngine.  `presolve` is now independent of the concern of the state since SmtEngine does it alone.